src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -275,30 +285,47 @@ private bool CData()
         {
             // Accept "/" and move next
             Assert(HtmlSymbolType.ForwardSlash);
-            var solidus = CurrentSymbol;
+            var forwardSlash = CurrentSymbol;
             if (!NextToken())
             {
                 Accept(_bufferedOpenAngle);
-                Accept(solidus);
+                Accept(forwardSlash);
                 return false;
             }
             else
             {
-                var tagName = String.Empty;
-                if (At(HtmlSymbolType.Text))
+                var tagName = string.Empty;
+                HtmlSymbol bangSymbol = null;
+
+                if (At(HtmlSymbolType.Bang))
+                {
+                    bangSymbol = CurrentSymbol;
+
+                    var nextSymbol = Lookahead();
+
+                    if (nextSymbol != null && nextSymbol.Type == HtmlSymbolType.Text)
+                    {
+                        tagName = "!" + nextSymbol.Content;
+                    }
+                }
+                else if (At(HtmlSymbolType.Text))
                 {
                     tagName = CurrentSymbol.Content;
                 }
+
                 var matched = RemoveTag(tags, tagName, tagStart);
 
                 if (tags.Count == 0 &&
-                    String.Equals(tagName, SyntaxConstants.TextTagName, StringComparison.OrdinalIgnoreCase) &&
+                    // Text tags cannot be targeted by TagHelpers and therefore can't be escaped with '!'.
```

Comment:
nit: on 3rd reading I'm not sure this comment adds much.  may be more useful to say `// Note tagName may contain a '!' escape character.  This ensures `</!text>` doesn't match here.`  might also be the place for code comments on what happens to `</!text>` after it isn't special-cased here.
- similar comment for the similar start tag code later in this file